codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttEncoder.java

```diff
@@ -689,10 +697,33 @@ private static void writeVariableLengthInt(ByteBuf buf, int num) {
         } while (num > 0);
     }
 
-    static void writeUTF8String(ByteBuf buf, String s) {
-        byte[] sBytes = encodeStringUtf8(s);
-        buf.writeShort(sBytes.length);
-        buf.writeBytes(sBytes, 0, sBytes.length);
+    private static int nullableUtf8Bytes(String s) {
+        if (s == null) {
+            return 0;
+        }
+        return utf8Bytes(s);
+    }
+
+    private static void writeUTF8String(ByteBuf buf, String s, int expectedUtf8Length) {
+        assert nullableUtf8Bytes(s) == expectedUtf8Length;
+        assert buf.isWritable(expectedUtf8Length + 2);
```

Comment:
This isn't necessary.. `buf` will be expanded inside `reserveAndWriteUtf8` if needed

Reply:
nope, because of the `short` length prefixed. That's why I've called `buf.ensureWritable(buf.capacity() - writerIndex);` explicitly